pallets/randomness/src/vrf.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! VRF logic
+use crate::{
+	Config, CurrentVrfInput, LocalVrfOutput, NotFirstBlock, RandomnessResults, RequestType,
+};
+use frame_support::{pallet_prelude::Weight, traits::Get};
+use nimbus_primitives::{NimbusId, NIMBUS_ENGINE_ID};
+use parity_scale_codec::Decode;
+pub use session_keys_primitives::make_transcript;
+use session_keys_primitives::{
+	GetVrfInput, KeysLookup, PreDigest, VrfId, VRF_ENGINE_ID, VRF_INOUT_CONTEXT,
+};
+use sp_consensus_vrf::schnorrkel;
+use sp_core::crypto::ByteArray;
+
+/// VRF output
+type Randomness = schnorrkel::Randomness;
+
+/// Set vrf input in storage and log warning if either of the values did NOT change
+/// Called in previous block's `on_finalize`
+pub(crate) fn set_input<T: Config>() {
+	let input = T::VrfInputGetter::get_vrf_input();
+	if let Some(last_vrf_input) = <CurrentVrfInput<T>>::take() {
+		// logs if input uniqueness assumptions are violated (no reuse of vrf inputs)
+		if last_vrf_input.storage_root == input.storage_root
+			|| last_vrf_input.slot_number == input.slot_number
+		{
+			log::warn!(
+				"VRF on_initialize: storage root or slot number did not change between \
+			current and last block. Nimbus would've panicked if slot number did not change \
+			so probably storage root did not change."
+			);
+		}
+	}
+	<CurrentVrfInput<T>>::put(input);
+}
+
+/// Returns weight consumed in `on_initialize`
+pub(crate) fn set_output<T: Config>() -> Weight {
```

Comment:
The fn name could describe itself a bit more, maybe `inject_randomness` or `inject_local_randomness`?